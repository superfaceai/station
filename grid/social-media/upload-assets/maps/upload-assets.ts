/* eslint-disable jest/no-export */
import {
  RecordingProcessOptions,
  SuperfaceTest,
  SuperfaceTestRun,
} from '@superfaceai/testing';

type ProfilesResult = Array<{ id: string; name: string }>;

export const getPublishingProfiles = async (
  provider: string
): Promise<ProfilesResult> => {
  const superfacePublishingProfiles = new SuperfaceTest({
    profile: 'social-media/publishing-profiles',
    provider,
  });
  const result = await superfacePublishingProfiles.run({
    useCase: 'GetProfilesForPublishing',
    input: {},
  });
  expect(result.isOk()).toBeTruthy();

  return (result.unwrap() as { profiles: ProfilesResult })?.profiles || [];
};

export const publishVideoTest = (
  provider: string,
  input: SuperfaceTestRun['input'],
  hooks?: RecordingProcessOptions
): void => {
  describe(`social-media/upload-assets/${provider}`, () => {
    let superfaceUploadAssets: SuperfaceTest;
    let superfacePublishPost: SuperfaceTest;

    beforeAll(() => {
      jest.setTimeout(60000);
    });

    afterAll(() => {
      jest.setTimeout(5000);
    });

    beforeEach(() => {
      superfaceUploadAssets = new SuperfaceTest({
        profile: 'social-media/upload-assets',
        provider,
      });
      superfacePublishPost = new SuperfaceTest({
        profile: 'social-media/publish-post',
        provider,
      });
    });

    let assetId: string;

    describe('UploadAssetViaUrl', () => {
      it('should return an id', async () => {
        const profiles = await getPublishingProfiles(provider);

        const result = await superfaceUploadAssets.run(
          {
            useCase: 'UploadAssetViaUrl',
            input: {
              profileId: profiles[0].id,
              ...input,
            },
          },
          hooks
        );

        expect(result.isOk()).toBe(true);
        expect(result).toMatchSnapshot();
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        assetId = (result.unwrap() as any).assetId as string;
      });
    });

    describe('GetAssetState', () => {
      test("should eventually return 'FINISHED'", () => {
        return new Promise<void>((resolve, reject) => {
          function checkForFinished() {
            superfaceUploadAssets
              .run(
                {
                  useCase: 'GetAssetState',
                  input: { assetId },
                },
                hooks
              )
              .then(result => {
                expect(result.isOk()).toBe(true);
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
                if ((result.unwrap() as any).state === 'Finished') {
                  resolve();
                } else {
                  setTimeout(checkForFinished, 5 * 1000);
                }
              })
              .catch(err => {
                reject(err);
              });
          }

          checkForFinished();
        });
      });
    });

    describe('Publish the Post', () => {
      test('video gets published', async () => {
        const profiles = await getPublishingProfiles(provider);

        const result = await superfacePublishPost.run(
          {
            useCase: 'PublishPost',
            input: {
              profileId: profiles[0].id,
              attachments: [{ id: assetId }],
            },
          },
          hooks
        );
        expect(result.isOk()).toBe(true);
        expect(result).toMatchSnapshot();
      });
    });
  });
};
