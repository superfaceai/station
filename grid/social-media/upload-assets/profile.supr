
"""
Upload Assets to Social Media sites
Upload images and videos on social media such as Instagram, Facebook, or Twitter.
"""

name = "social-media/upload-assets"
version = "1.0.0"

"""
Upload an Asset
Register an upload of an asset to a social media profile, in preparation for publishing later.
"""
usecase UploadAssetViaUrl unsafe {
  input {
    """
    Profile ID
    Identifier of a profile to upload to. May be optional with some providers.
    """
    profileId string!

    """
    Caption
    Caption of the asset
    """
    caption string!

    """
    URL
    URL of the asset. Must be publically acessible.
    """
    url! string!

    """
    Type
    Media type of the asset
    """
    type enum {
      video,
      image
    }
  }

  result {
    """
    Asset ID
    Identifier of the asset.
    """
    assetId! string!
  }

  error {
    """
    Title
    A short, human-readable summary of the problem type.
    """
    title! string

    """
    Detail
    A human-readable explanation specific to this occurrence of the problem.
    """
    detail string

    """
    Code
    Normalized error code usable across providers
    """
    code ErrorCode

    """
    Original Error
    Original response from the provider
    """
    original
  }

  example Successful {
    input {
      profileId = '110658944785122',
      text = 'A video of a kitten falling over.',
      url = 'https://example.com/kitten.mp4',
    }

    result {
      assetId = '125682506616099',
    }
  }

  example Failed {
    input {
      profileId = 'invalid profile ID',
      caption = 'Caption of the thing.',
    }

    error {
      title = 'Bad request'
    }
  }
}

"""
Get Asset State
Get the state of a previously registered asset.
"""
usecase GetAssetState safe {
  input {
    """
    Asset ID
    Identifier of the asset.
    """
    assetId! string!
  }

  result {
    """
    State
    State of the asset. Can be one of the following:
      - Expired (Asset did not get published in time)
      - Error (Error during processing of asset)
      - Finished (Asset is ready to be published)
      - InProgress (Asset is currently being processed)
      - Published (Asset has been successfuly published)
    """
    state enum {
      Expired,
      Error,
      Finished,
      InProgress,
      Published
    }
  }

error {
    """
    Title
    A short, human-readable summary of the problem type.
    """
    title! string

    """
    Detail
    A human-readable explanation specific to this occurrence of the problem.
    """
    detail string

    """
    Code
    Normalized error code usable across providers
    """
    code ErrorCode

    """
    Original Error
    Original response from the provider
    """
    original
  }

  example Successful {
    input {
      assetId = '125682506616099',
    }

    result {
      state = 'InProgress',
    }
  }

  example Failed {
    input {
      profileId = 'invalid asset ID',
    }

    error {
      title = 'Bad request'
    }
  }
}

model ErrorCode enum {
  """
  Bad request
  The request was invalid or cannot be otherwise served.
  """
  BadRequest
  """
  Not found
  The profileId, container or other input item does not exist.
  """
  NotFound
  """
  Rate limit reached
  Returned when a request cannot be served due to the user or app reached max requests quota.
  """
  RateLimitReached
  """
  Unauthenticated
  There was a problem authenticating your request. Check that you are passing valid access token.
  """
  Unauthenticated
  """
  Unauthorized
  The request is understood, but it has been refused or access is not allowed.
  """
  Unauthorized

  Timeout
  ServerError
  Spam
  RestrictedProfile
  PayloadTooLarge
  UnsupportedAttachmentFormat

  """
  Unknown error
  Returned when an unexpected error occured.
  """
  UnknownError
}
