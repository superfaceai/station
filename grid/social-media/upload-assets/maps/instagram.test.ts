import { publishVideoTest } from './upload-assets';

publishVideoTest('instagram', {
  caption: 'Single video from Superface Station.',
  url: 'https://download.samplelib.com/mp4/sample-5s.mp4',
});
