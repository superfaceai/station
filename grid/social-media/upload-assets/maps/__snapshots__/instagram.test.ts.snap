// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`social-media/upload-assets/instagram Publish the Post video gets published 1`] = `
Ok {
  "value": Object {
    "postId": "17934169565396380",
    "url": "https://www.instagram.com/p/CiLMrsEqAYE/",
  },
}
`;

exports[`social-media/upload-assets/instagram UploadAssetViaUrl should return an id 1`] = `
Ok {
  "value": Object {
    "assetId": "17985138253529188",
  },
}
`;
